File: LeetCode/0025.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        if (k == 1 || head == nullptr)
            return head;
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode *pre = dummy, *cur = head;
        int cnt = 0;
        while (cur != nullptr)
        {
            cnt++;
            cur = cur->next;
        }
        cur = head;
        for (int i = 0; i < cnt / k; i++)
        {
            for (int j = 1; j < k; j++)
            {
                ListNode *next = cur->next;
                cur->next = next->next;
                next->next = pre->next;
                pre->next = next;
            }
            pre = cur;
            cur = cur->next;
        }
        return dummy->next;
    }
};

File: leetcode/0025.go
package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	if k == 1 || head == nil {
		return head
	}

	dummynode := &ListNode{Val: -1}
	dummynode.Next = head
	pre, cur, cnt := dummynode, head, 0

	for cur != nil {
		cnt++
		cur = cur.Next
	}
	cur = head
	for i := 0; i < cnt/k; i++ {
		for j := 1; j < k; j++ {
			next := cur.Next
			cur.Next = next.Next
			next.Next = pre.Next
			pre.Next = next
		}
		pre = cur
		cur = cur.Next
	}
	return dummynode.Next
}
